File: daily-topic/nextPermutation.go
package daily_topic

func nextPermutation(nums []int) {
	for i := len(nums) - 2; i >= 0; i-- {
		min := i
		for j := len(nums) - 1; j > i; j-- {
			if (nums[j] > nums[i]) && (min == i || nums[j] < nums[min]) {
				min = j
			}
		}
		if min != i {
			tmp := nums[min]
			for k := min; k > i; k-- {
				nums[k] = nums[k-1]
			}
			nums[i] = tmp
			return
		}
	}
	for i := 0; i < len(nums)/2; i++ {
		t := nums[i]
		nums[i] = nums[len(nums)-1-i]
		nums[len(nums)-1-i] = t
	}
}

File: host100/NextPermutation.java
import java.util.Arrays;
import java.util.Stack;

/**
 * @author : zexho
 * @created : 2022-06-20
**/
public class NextPermutation{
    public void nextPermutation(int[] nums) {
        int len = nums.length;
        Stack<Integer> stack = new Stack<Integer>();
        stack.push(len-1);

        int i = nums.length-2;
        int idx = -1;
        while(i >= 0){
            if (nums[i] < nums[stack.peek()]) {
                while(!stack.isEmpty() && nums[i] < nums[stack.peek()]){
                    idx = stack.pop();
                }
                int t = nums[idx];
                nums[idx] = nums[i];
                nums[i] = t;
                break;
            }
            stack.push(i);
            i--;
        }
        Arrays.sort(nums,i+1,len);
    }

}
